discovery_engine_core/core/src/mab.rs

```diff
@@ -46,113 +58,115 @@ impl BetaSample for BetaSampler {
+                            Some((sample, index))
+                        } else {
+                            max
+                        }
+                    } else {
+                        Some((sample, index))
+                    }
+                })
+        }) {
+        Ok(Some((_, index))) => Some(Ok(index)),
+        Ok(None) => None,
+        Err(error) => Some(Err(error)),
+    }
 }
 
-struct SelectionIter<'bs, 'b, BS, B, T>
-where
-    B: Bucket<T>,
-{
-    beta_sampler: &'bs BS,
-    buckets: Vec<&'b mut B>,
-    bucket_type: PhantomData<T>,
+/// A stream to select elements from buckets.
+pub(crate) struct Selection<'b, BS, B, T> {
+    beta_sampler: BS,
+    buckets: Vec<&'b RwLock<B>>,
+    bucket_type: PhantomData<&'b T>,
 }
 
-impl<'bs, 'b, BS, B, T> SelectionIter<'bs, 'b, BS, B, T>
-where
-    B: Bucket<T>,
-{
-    fn new(beta_sampler: &'bs BS, buckets: Vec<&'b mut B>) -> Self {
+impl<'b, BS, B, T> Selection<'b, BS, B, T> {
+    /// Creates a selective steam.
+    pub(crate) fn new(beta_sampler: BS, buckets: impl IntoIterator<Item = &'b RwLock<B>>) -> Self {
         Self {
             beta_sampler,
-            buckets,
+            buckets: buckets.into_iter().collect(),
             bucket_type: PhantomData,
         }
     }
+
+    /// Selects up to n elements.
+    #[must_use = "futures do nothing unless they are awaited/polled"]
+    pub(crate) fn select(
+        self,
+        n: usize,
+    ) -> impl 'b + Future<Output = Result<Vec<T>, Error>> + Send + Sync
+    where
+        BS: 'b + BetaSample + Send + Sync,
+        B: Bucket<T> + Send + Sync,
+        T: Send + Sync,
+    {
+        self.take(n).try_collect()
+    }
 }
 
-impl<'bs, 'b, BS, B, T> Iterator for SelectionIter<'bs, 'b, BS, B, T>
+impl<'b, BS, B, T> Stream for Selection<'b, BS, B, T>
 where
     BS: BetaSample,
     B: Bucket<T>,
 {
     type Item = Result<T, Error>;
 
-    fn next(&mut self) -> Option<Self::Item> {
-        let mut buckets = vec![];
-        std::mem::swap(&mut self.buckets, &mut buckets);
-
-        self.buckets = buckets
-            .into_iter()
-            .filter(|bucket| !bucket.is_empty())
-            .collect::<Vec<&mut B>>();
-
-        if self.buckets.is_empty() {
-            None
-        } else {
-            Some(pull_arms(self.beta_sampler, &mut self.buckets))
+    fn poll_next(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
+        let buckets = self
+            .buckets
+            .iter()
+            .map(|bucket| bucket.read())
+            .collect::<FuturesUnordered<_>>()
+            .collect::<Vec<_>>();
+        pin_mut!(buckets);
+
+        match buckets
+            .poll(cx)
+            .map(|buckets| pull_arms(&self.beta_sampler, buckets))
+        {
+            Poll::Ready(Some(Ok(index))) => {
+                let bucket = self.buckets[index].write();
+                pin_mut!(bucket);
```

Comment:
Same as the other case it should be fine the way we use it.

But if this could be run concurrent there could be minor "off" results if the second run starts between releasing the read lock and gaining the write lock.

Reply:
This does not solve the original problem. When ask for k elements we want to extract them from the current state we no update in between.
Seems that we cannot download new items from the stack and get items from them at the same time with this approach, we might need to separate the operations from the data.

Reply:
i'll rewrite the code in a way that we acquire all write locks for all stacks (instead of all read locks and afterwards one write lock) and pop from the sampled stack before releasing any lock. this way we guarantee that the state hasn't changed, but on the other hand it is more restrictive of course (in the current context this should be ok).